src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSA.cs

```diff
@@ -980,6 +1362,79 @@ private static void ClearPrivateParameters(in RSAParameters rsaParameters)
+            // deterministic, so start with the key size.
+            int resultSize = (KeySize + 7) / 8;
+            int written;
+
+            // For scenarios where we are confident that we can get the output side right on the first try, we allocate
+            // and use that as the buffer. This is the case for signing and encryption since that is always going to be
+            // the modulus size.
+            // For decryption, we go straight to renting as there is no way to know the size of the final output prior
+            // to decryption, so the allocation would probably be wasted.
+            if (tryKeySizeFirst)
+            {
+                byte[] result = new byte[resultSize];
+
+                if (callback(this, input, result, state, out written))
+                {
+                    if (written <= resultSize)
+                    {
+                        // In a typical sign or encrypt scenario, Resize won't do anything so we return the array as-is.
+                        // On the offchance some implementation returns less than what we expected, shrink the result.
+                        Array.Resize(ref result, written);
+                        return result;
+                    }
+
+                    // The Try virtual returned a bogus value for written. It returned a written value larger
+                    // than the buffer passed in. This means the Try implementation is not reliable, so throw
+                    // the best exception we can.
+                    throw new CryptographicException(SR.Argument_DestinationTooShort);
+                }
+
+                // We're about to try renting from the pool, so the next rental should be bigger than what we just tried.
+                resultSize = checked(resultSize * 2);
+            }
+
+            while (true)
+            {
+                // Use ArrayPool instead of CryptoPool since we are handing this out to a virtual.
+                byte[] rented = ArrayPool<byte>.Shared.Rent(resultSize);
+                byte[]? rentResult = null;
+
+                if (callback(this, input, rented, state, out written))
+                {
+                    if (written > rented.Length)
+                    {
+                        // The virtual did something unexpected, so don't return the array to the pool.
+                        // Consistency throw the same exception if the Try method wrote an impossible amount of data.
+                        throw new CryptographicException(SR.Argument_DestinationTooShort);
+                    }
+
+                    rentResult = rented.AsSpan(0, written).ToArray();
+                }
+
+                CryptographicOperations.ZeroMemory(rented.AsSpan(0, written));
+                ArrayPool<byte>.Shared.Return(rented);
+
+                if (rentResult is not null)
+                {
+                    return rentResult;
+                }
+
+                resultSize = checked(resultSize * 2);
```

Comment:
Similarly, what does growth mean here?  That decrypt has returned an answer bigger than the key, which isn't possible (or, at least, you've stopped being RSA at that point).

It feels like we have "I expect the array, let's just allocate it first" and "I expect smaller than the array, let's rent it", and either way it's one and done.  No?